### DISCUSSION TITLE
Decoration is not always drawn.

### DISCUSSION THREAD

**Discussion by peshkira** [patrykandpatrick/vico#918] [Questions] [answered]
### Question

I have a simple Now Label decoration, which I want to display whenever the user scrolls to the end of the chart and hide if the user scrolls away. The chart actually increases its bounds each minute so that the end is always at "Now".

For the most part it works as intended, but sometimes when I scroll with a fling or just faster, the canvas is not redrawn and the decoration does not appear until the next draw is scheduled (e.g. when new data comes in, or another wall-clock minute passes).

The state seems to be updated correctly based on the scroll position and I see that it actually changes the list of decorations.
Here is the relevant portion of the code, I guess.
I'd appreciate it, if you can give me a hint to what I might be doing wrong or if there is a way to force the chart to redraw, without actually changing the data set.

Thanks.

```
with(chartState) {
    CartesianChartHost(
        chart = rememberCartesianChart(
            layers = rememberLineCartesianLayers(),
            endAxis = rememberEndAxis(),
            decorations = rememberDecorations(),
            persistentMarkers = rememberPersistentMarkers()
        ),
        modelProducer = modelProducer,
        zoomState = zoomState,
        scrollState = scrollState,
        modifier = modifier.fillMaxHeight()
    )
}
```

The decorations extension

```
@Composable
private fun ChartState.rememberDecorations(): List<Decoration> {
    val chartState = this
    return buildList {
        // add other permanent decorations, which work as intended

        if (chartState.nowLabelVisible) {
            add(rememberNowLabel())
        }
    }
}
```

```
@Composable
private fun rememberNowLabel(): NowLabel {
    return rememberNowLabel(
        label = stringResource(id = ...),
        labelComponent = rememberAxisLabelComponent(color = White)
    )
}
```

```
public class NowLabel(
    private val label: String,
    private val labelComponent: TextComponent? = null,
) : Decoration {

    override fun drawOverLayers(context: CartesianDrawingContext) {
        with(context) {
            labelComponent?.draw(
                this,
                text = label,
                x = layerBounds.getEnd(isLtr),
                y = layerBounds.bottom + 50
            )
        }
    }
}

@Composable
public fun rememberNowLabel(
    label: String,
    labelComponent: TextComponent?
): NowLabel = remember(label, labelComponent) {
    NowLabel(
        label = label,
        labelComponent = labelComponent
    )
}
```

### Vico version(s)

2.0.0.beta1

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @peshkira. Please provide a complete snippet that demonstrates the issue. We have no way of reproducing it using the code you've shared, as the logic that shows and hides the custom `Decoration` is missing.

**Reply by peshkira** [reply]
I just have a onScroll lambda passed into the composable that handles the chart
I then have a LaunchedEffect, which subscribes to the scrollState and calls the lambda.

The parent composable dispatches an action to the viewModel, which would change the state (a parameter to the chart) you see in the snippets above `if (chartState.nowLabelVisible)`

```
 LaunchedEffect(scrollState) {
        snapshotFlow { scrollState.value }
            .distinctUntilChanged()
            .collect {         
                onScroll(scrollState.maxValue - it <= 0.1)
            }
    }
```

I can try to provide a full working example later, but I need to migrate it to a separate project.

**Reply by Gowsky** [reply]
> I just have a onScroll lambda passed into the composable that handles the chart
> I then have a LaunchedEffect, which subscribes to the scrollState and calls the lambda.
> 
> The parent composable dispatches an action to the viewModel, which would change the state (a parameter to the chart) you see in the snippets above if (chartState.nowLabelVisible)

This is helpful context, but it's still a puzzle to piece everything together and try to replicate the issue accurately.

> I can try to provide a full working example later

Please do. We can then help more effectively.

> but I need to migrate it to a separate project

A full project isn't required. We just need a self-contained snippet with a composable we can run. See #659 for an example.

**Reply by peshkira** [reply]
Hey @Gowsky!

I just wanted to sit down and extract a small working example and I noticed you released beta.2. I tried it and it seems to fix the problem.

My suspicion was that the recomposition does not happen and thus a call to draw is not forced when the decorations change.
Now it seems to work every time consistently exactly as I expect it to. So, I assume the problem was related.

Thanks! I will close this discussion as fixed with 2.0.0-beta.2

**Comment by peshkira** [comment, marked as answer]
Fixed in 2.0.0-beta.2
